[Merged by Bors] - feat(Mathlib/CategoryTheory/EffectiveEpi/Types): Effective epi in types

yonggyuchoimath opened the pull request:
This PR characterizes effective epimorphisms in the category `Type u`, proving that a function of types is an effective epimorphism if and only if it is surjective.

The main results:

`regularEpi_iff_surjective`: A function f : `X → Y` in `Type u` is a regular epimorphism if and only if it is surjective.
`effectiveEpi_iff_surjective`: A function f : `X → Y` in `Type u` is an effective epimorphism if and only if it is surjective.

yonggyuchoimath committed:
effective epi in types

yonggyuchoimath committed:
lake exe mk_all

joelriou commented on Mathlib/CategoryTheory/EffectiveEpi/Types.lean:
```diff
@@ -0,0 +1,113 @@
+  obtain ⟨x, _⟩ := hf y
+  exact ⟨Quot.mk _ x, by simpa⟩
+
+/-- If `f : X ⟶ Y` is surjective, then `fromCoequalizerPullbackCone f` is surjective. -/
```
Note that we do not require docstrings for lemmas, especially when the statement is very clear from the code, which is the case here:
```suggestion
/-- If `f : X ⟶ Y` is surjective, then `fromCoequalizerPullbackCone f` is bijective. -/
```

yonggyuchoimath committed:
Update Mathlib/CategoryTheory/EffectiveEpi/Types.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

yonggyuchoimath committed:
Update Mathlib/CategoryTheory/EffectiveEpi/Types.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

yonggyuchoimath committed:
Update Mathlib/CategoryTheory/EffectiveEpi/Types.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

yonggyuchoimath commented:
Thanks for your review!

joelriou commented on Mathlib/CategoryTheory/EffectiveEpi/Types.lean:
```diff
@@ -0,0 +1,113 @@
+  right := (Types.pullbackCone f f).snd
+  w := funext fun p => p.2
+  isColimit := by
+    refine Cofork.isColimitOfIsos _
+      (coequalizerIsCoequalizer (Types.pullbackCone f f).fst (Types.pullbackCone f f).snd) _
+      (.refl _) (.refl _) (coequalizerPullbackConeOfSurjective hf) ?_ ?_ ?_
+    · simp
+    · simp
+    · rw [coequalizerPullbackConeOfSurjective]
+      cat_disch
```
Untested:
```suggestion
    refine Cofork.isColimitOfIsos _
      (coequalizerIsCoequalizer (Types.pullbackCone f f).fst (Types.pullbackCone f f).snd) _
      (.refl _) (.refl _) (coequalizerPullbackConeOfSurjective hf) (by simp) (by simp) ?_
    rw [coequalizerPullbackConeOfSurjective]
    cat_disch
```

yonggyuchoimath committed:
Update Mathlib/CategoryTheory/EffectiveEpi/Types.lean

Co-authored-by: Joël Riou <37772949+joelriou@users.noreply.github.com>

yonggyuchoimath commented on Mathlib/CategoryTheory/EffectiveEpi/Types.lean:
```diff
@@ -0,0 +1,113 @@
+  intro z₁ z₂ h
+  rw [← Quot.out_eq z₁, ← Quot.out_eq z₂]
+  have h' : f (Quot.out z₁) = f (Quot.out z₂) := by
+    rwa [← Coequalizer.desc_mk _ _ f _ _, ← Coequalizer.desc_mk _ _ f _ _,
+      Coequalizer.mk, Coequalizer.mk, Quot.out_eq, Quot.out_eq]
```
Changing them to `← Coequalizer.desc_mk _ _ f` works, but `← Coequalizer.desc_mk` returns:
```
Invalid rewrite argument: The pattern to be substituted is a metavariable (`?u ?x`) in this equality
  ?u ?x = Coequalizer.desc ?f ?g ?u ?hu (Coequalizer.mk ?f ?g ?x)
```

-awaiting-author

joelriou replied:
Then, please remove the last two `_`?

yonggyuchoimath committed:
remove last _ _

yonggyuchoimath commented:
-awaiting-author

erdOne reviewed (commented):
I have yet to read contents of the PR in full but I think we already have the result mentioned in the PR comment via [`CategoryTheory.RegularEpi.ofSplitEpi`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Limits/Shapes/RegularMono.html#CategoryTheory.RegularEpi.ofSplitEpi) and [`CategoryTheory.instEffectiveEpiOfRegularEpi`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/EffectiveEpi/RegularEpi.html#CategoryTheory.instEffectiveEpiOfRegularEpi) and [`CategoryTheory.instSplitEpiCategoryType`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/CategoryTheory/Types/Basic.html#CategoryTheory.instSplitEpiCategoryType)?

In particular this works in current mathlib
```
import Mathlib

open CategoryTheory

attribute [instance] isSplitEpi_of_epi

instance {X Y : Type _} (f : X ⟶ Y) [Epi f] : EffectiveEpi f := inferInstance
```

erdOne commented:
Is `CategoryTheory.regularEpiOfEffectiveEpi` what you want?

yonggyuchoimath commented:
> regularEpiOfEffectiveEpi

Thank you for pointing that out! It seems that I actually already used it under `infer_instance` in the PR...

Should I close this PR?

erdOne commented:
Maybe you can still add `Surjective f -> IsSplitEpi f` and `Surjective f -> EffectiveEpi f` somewhere if you find them useful?

yonggyuchoimath committed:
remove file and add isSplitEpi_iff_surjective

yonggyuchoimath commented:
I've removed the `Mathlib.CategoryTheory.EffectiveEpi.Types.lean`, and instead implemented `isSplitEpi_iff_surjective` in `Mathlib.CategoryTheory.Types.Basic.lean`. Additionally, I removed one unnecessary import.

I didn't find a suitable location to add `Surjective f -> EffectiveEpi f` or `Surjective f -> RegularEpi f`, since adding these to `Mathlib.CategoryTheory.Types.Basic.lean` would introduce circular imports. Anyway, once I import `Mathlib.CategoryTheory.EffectiveEpi.RegularEpi` and use `have := (isSplitEpi_iff_surjective f).mpr ‹_›; inferInstance`, I was able to get both `RegularEpi f` and `EffectiveEpi f`.

I don't know how topos theory is being implemented in Mathlib 4; there may be a more appropriate place to include these things? I would appreciate any suggestions on this.

yonggyuchoimath committed:
avoid @

yonggyuchoimath commented on Mathlib/CategoryTheory/Types/Basic.lean:
```diff
@@ -358,6 +357,11 @@ instance : SplitEpiCategory (Type u) where
+theorem isSplitEpi_iff_surjective {X Y : Type u} (f : X ⟶ Y) :
+    IsSplitEpi f ↔ Function.Surjective f :=
+  Iff.intro (fun _ => surjective_of_epi _)
+    fun hf => @isSplitEpi_of_epi _ _ _ _ _ _ ((epi_iff_surjective f).mpr hf)
```
I've changed it into
```
theorem isSplitEpi_iff_surjective {X Y : Type u} (f : X ⟶ Y) :
    IsSplitEpi f ↔ Function.Surjective f :=
  Iff.intro (fun _ => surjective_of_epi _)
    fun hf => (by simp only [(epi_iff_surjective f).mpr hf, isSplitEpi_of_epi])
```
How does this look?

yonggyuchoimath commented:
-awaiting-author
